### PR TITLE
Sync: ensure we can calculate a histogram for term_taxonomy

### PR DESCRIPTION
**Pull request by roccotripaldi**
Fixes an issue where we cannot validate a site where term_taxonomy is out of sync

#### Changes proposed in this Pull Request:

* Ensure sync histogram endpoint can calculate a checksum for term_taxonomy

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Ip7rcWF-12N-p2

#### Testing instructions:
* requires an upstream patch D30489-code to test

#### Proposed changelog entry for your changes:

* None

### CODE REVIEW

**Review by tyxla — DISMISSED** [review]
Hm, it feels odd to me that we're treating term taxonomy table as a table that is responsible for a separate entity. Terms and term taxonomies are two sides of the same thing, are architectured separately because of historical reasons, and they're working together, so I think we should be syncing them (and validating/fixing them) together.

Also, we're currently syncing both tables in the Terms module (expanding terms in `expand_term_taxonomy_id` will result in having `WP_Term` objects, which uses both tables), so why do we need to have it as a separate sync module?

I think that instead we should try to have a universal checksum for terms, one that considers both tables. After all, if there is a difference in one of the tables, that essentially affects the other table as well (we can't be sure that the other table is valid anymore).

And if we want to have a separate histogram for the term taxonomy table, it feels odd to sync that table in another sync module.

What do you think?

**Comment by roccotripaldi** [comment]
great feed, @tyxla - i think i can make a change here where we don't need a separate module.

**Comment by roccotripaldi** [comment]
Note: some commits are not verified because we are still working, in separate PRs, on quelling the phpcs in the Sync package.

File: packages/sync/src/modules/Terms.php [hunk 1]
```diff
@@ -35,29 +35,34 @@ public function name() {
...
 
 	/**
-	 * Allows WordPress.com servers to retrieve a term object via the sync API.
+	 * Allows WordPress.com servers to retrieve term-related objects via the sync API.
 	 *
 	 * @param string $object_type The type of object.
 	 * @param int    $id          The id of the object.
 	 *
-	 * @return bool|\WP_Term
+	 * @return bool|object A WP_Term object, or a row from term_taxonomy table based on
```
Comment (tyxla) [inline packages/sync/src/modules/Terms.php]: Seems like we need to add to the description here.

File: packages/sync/src/modules/Terms.php [hunk 2]
```diff
@@ -35,29 +35,34 @@ public function name() {
...
 	 */
 	public function get_object_by_id( $object_type, $id ) {
+		global $wpdb;
+		$object = false;
 		if ( 'term' === $object_type ) {
-			$term = get_term( intval( $id ) );
+			$object = get_term( intval( $id ) );
```
Comment (tyxla) [inline packages/sync/src/modules/Terms.php]: Something that is still weird here is that we still use `get_term()` for `term` objects, and that includes parts that pull from the term taxonomy table; while the `term_taxonomy` object below is a raw db row pull of a particular row. Should this one also work with raw term rows too?

**Review by tyxla — COMMENTED** [review]
Thanks Rocco, this is great work!

I've left a couple of questions, mainly related with the overlap that still exists between terms and term taxonomy, in contrast of the separation we're trying to achieve when calculating histograms.

File: packages/sync/src/modules/Terms.php (hunk 2 above)
Comment (roccotripaldi) [reply packages/sync/src/modules/Terms.php]: great call!

Comment (roccotripaldi) [reply packages/sync/src/modules/Terms.php]: On second though, on the server side, we will only have access to term_taxonomy_id when we calculate the checksum for term_taxonomy.

**Review by gititon — DISMISSED** [review]
Looks like all of the prior feedback was addressed, code looks good!

**Comment by jeherve** [comment]
I tried testing this following the instructions in D30489-code, but the validator script never seems to finish in my tests. Are there other tests I should be running?

**Review by lezama — APPROVED** [review]

**Comment by lezama** [comment]
> I think that instead we should try to have a universal checksum for terms, one that considers both tables

@tyxla we opted to not do a join between the two tables for performance reasons but we might need to revisit this.

**Pull request merged**